alibi/models/pytorch/tests/test_model.py

```diff
@@ -0,0 +1,618 @@
+    y_true = dataset['output']['multimodal']['y_true']
+
+    with reset_model(multimodal_model):
+        results = multimodal_model.compute_metrics(y_pred=y_pred, y_true=y_true)
+
+    expected_acc1 = torch.mean((y_true[0] == torch.argmax(y_pred[0], dim=-1)).float()).item()
+    expected_acc2 = torch.mean((y_true[1] == torch.argmax(y_pred[1], dim=-1)).float()).item()
+    assert np.isclose(expected_acc1, results['output_1_accuracy'])
+    assert np.isclose(expected_acc2, results['output_2_accuracy'])
+
+
+@pytest.mark.parametrize('dataset', ['regression'], indirect=True)
+@pytest.mark.parametrize('unimodal_model', [
+    {
+        'input_dim': input_dim,
+        'output_dim': output_dim,
+        'compile': True,
+        'optimizer': optimizer_class,
+        'loss': nn.MSELoss(reduction='mean'),
+    }
+], indirect=True)
+def test_train_step(unimodal_model, dataset):
+    """ Test if the train step return the appropriate statistics. """
+    # copy the model `state_dict` as it will be modified by the `train_step`. Not really necessary now,
+    # but to avoid any errors in the future if it is the case.
+    state_dict_cpy = deepcopy(unimodal_model.state_dict())
+
+    x = dataset['input']
+    y_true = dataset['output']['unimodal']['y_true']
+
+    # compute the loss manually
+    w = unimodal_model.fc1.weight.clone()
+    diff = y_true - x @ w.T
+    expected_loss = (torch.trace(diff @ diff.T) / (n_instances * output_dim)).item()
+
+    # compute the gradients manually
+    grad = -2 * (diff.T @ x) / (n_instances * output_dim)
+    new_w = w - lr * grad
+
+    # perform training step
+    with reset_model(unimodal_model):
+        results = unimodal_model.train_step(x, y_true)
+
+    # check if the loss and the updated weights match the manually computed ones
+    assert np.isclose(results['loss'], expected_loss)
+    assert torch.allclose(unimodal_model.fc1.weight, new_w)
+
+    # restore old `state_dict`
+    unimodal_model.load_state_dict(state_dict_cpy)
+
+
+@pytest.mark.parametrize('dataset', ['classification'], indirect=True)
+@pytest.mark.parametrize('unimodal_model', [
+    {
+        'input_dim': input_dim,
+        'output_dim': output_dim,
+        'compile': True,
+        'optimizer': optimizer_class,
+        'loss': loss_fn,
+        'metrics': metric,
+    }
+], indirect=True)
+def test_test_step(unimodal_model, dataset):
+    """ Test if the test step return the expected statistics. """
+    x = dataset['input']
+    y_true = dataset['output']['unimodal']['y_true']
+
+    # reset gradients
+    unimodal_model.optimizer.zero_grad()
+
+    # perform test step
+    with reset_model(unimodal_model):
+        results = unimodal_model.test_step(x, y_true)
+
+    # check if gradients are not computed
+    for param in unimodal_model.parameters():
+        assert (param.grad is None) or (torch.allclose(param, torch.zeros_like(param)))
+
+    # compute prediction
+    unimodal_model.eval()
```

Comment:
Should this be reverted before exiting the test (also see previous comment on fixtures)?

Reply:
Doesn't really matter for those tests since the model has the same behavior in train and eval. Now that I think about it, probably it makes sense to set the scope to function level ...